### PR TITLE
E2e/e2e merchant orders add order complex

### LINKED ISSUE #29003: Write new e2e test: Merchant – Orders – Add new Order complex – multiple product types & tax classes
Write a new e2e test that covers the following flow: 

- Merchant / Orders / Add new Order complex (multiple product types & tax classes)

### PR DESCRIPTION
**Pull request by rodelgc**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This Pull Request will add an e2e test that would cover the `Merchant | Orders | Add new Order complex (multiple product types & tax classes)` critical flow.
Closes #29003.
Below are the steps that this test does:

#### Test Data
1. **Products** - this test will create and use 4 products as test data:
    - One simple product named `Simple Product 273722`
    - One variable product named `Variable Product 024611`
    - One grouped product named `Grouped Product 858012`
    - One external product named `External product 786794`

    Since the test suite creates several products having the same names specified in `default.json`, I made this test's product names unique by suffixing them with a 6-digit random number, like `Simple Product 273722` so that the test would be able to select them correctly.

    I also did not follow the default `regularPrice` in `default.json` but instead specified my own, so that this test wouldn't fail when `default.json` is later modified.

2. **Tax Classes** - this test will create and use 3 tax classes:
    - One tax class for the simple product
    - One tax class for the variable product
    - One tax class for the external product

    I'm hesitant to use the default tax classes (Standard, Zero Rate, Reduced Rate, etc.) because I'm afraid that any changes to them by other tests in the future will break this test.

3. **Tax Rates** - this test will create and use 3 tax rates, one for each of the tax classes mentioned above.

#### Pre-test Setup (`beforeAll` block)
1. Using the API:
    1. Toggle on the "Enable tax rates and calculations" setting. Even though the `activate-and-setup` tests will turn this setting on, I still specified it here to be sure.
    2. Create the tax classes, tax rates, and products I mentioned above in the _"Test Data"_ section.
2. Using the UI, login as the merchant.
    
#### Test steps
1. Go to "add order" page.
2. Click 'Add item(s)' button.
3. Click 'Add Product(s)' button
4. In the modal that appears, type search for the following products and click on them in the results menu:
    - Simple Product 273722
    - Variable Product 024611
    - Grouped Product 858012
    - External product 786794
5. Click the 'Add' button in the modal window
6. Click 'Recalculate' button.
7. Click 'Create' button.
8. Verify that the table shows all the 4 product names.
9. Verify that the table headers show all the names of the 4 tax classes.
10. Verify that the table shows the expected tax rates.
11. Verify that the order totals below the table contain the correct tax names and amounts.

At the end of the test, it would have created an order containing the following products and taxes:
HN127xT4xs

### How to test the changes in this Pull Request:

1. After running `docker:up`, run the following commands:
```
# set initial state of the test
npx wc-e2e test:e2e tests/e2e/specs/activate-and-setup/setup-onboarding.test.js

# run this test
npx wc-e2e test:e2e tests/e2e/specs/wp-admin/test-create-order.js
```
2. Wait for the test to finish. You should see a passing `can create new complex order with multiple product types & tax classes` test
Ehl7g3opZB

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

### CODE REVIEW

**Comment by rrennick** [comment]
@rodel-calasagsag When a test requires modifying any of the existing entries in the file you need to verify that your changes do not break other tests. The tax classes don't exist for the other tests that use the entries in `default.json` to create products:
```
 FAIL  ../specs/rest-api/api.test.js (8.603s)
  REST API > External Product
    ✕ can create an external product (173ms)
    ✓ can retrieve a raw external product (75ms)
    ✕ can retrieve a transformed external product (76ms)
    ✓ can delete an external product (3ms)
....
  REST API > Variable Product
    ✕ can create a variable product (260ms)
    ✓ can add variations (509ms)
    ✕ can retrieve a transformed variable product (87ms)
    ✓ can retrieve transformed product variations (77ms)
    ✓ can delete a variation (3ms)
    ✓ can delete a variable product (2ms)
```

Since the tax classes are unique to this test, you should add them to the products programmatically instead of adding to the default configuration.

The `core-tests` folder is a [npm package](https://www.npmjs.com/package/@woocommerce/e2e-core-tests) that uses `tests/e2e/config/default.json` in both this repository and in repositories that use it as a package. When adding or changing entries in `default.json`, the entries do not get automatically added to the downstream repositories. Add fallback code to the test to avoid it immediately failing downstream where the data isn't present. (eg. https://github.com/woocommerce/woocommerce/blob/trunk/tests/e2e/core-tests/specs/merchant/wp-admin-product-new.test.js#L28)

**Comment by rodel-calasagsag** [comment]
Thanks @rrennick
I'll make the necesseary changes

File: tests/e2e/core-tests/specs/merchant/wp-admin-order-new.test.js [hunk 1]
```diff
@@ -4,12 +4,196 @@
...
+		class: 'tax-class-variable'
+	},
+	{
+		name: 'Tax Rate External',
+		rate: '30',
+		class: 'tax-class-external'
+	}
+];
+
+const taxTotals = ['£10.00', '£40.00', '£240.00'];
```
Comment (rrennick) [inline tests/e2e/core-tests/specs/merchant/wp-admin-order-new.test.js:38]: @rodel-calasagsag The currency used in the e2e tests is USD. You can initialize your test environment to a consistent state with `npx wc-e2e test:e2e tests/e2e/specs/activate-and-setup/setup-onboarding.test.js` after doing `docker:up`.

**Comment by tammullen** [comment]
Thanks @rodel-calasagsag!

Can you expand the _Changes proposed in this Pull Request:_ section to include some lower level details of the flow which are being covered in this test like the contents of the order and any other relevant details.

File: tests/e2e/core-tests/specs/merchant/wp-admin-order-new.test.js (hunk 1 above)
Comment (rodel-calasagsag) [reply tests/e2e/core-tests/specs/merchant/wp-admin-order-new.test.js:38]: Got it. Thanks @rrennick. I'll do that.

Comment (rodel-calasagsag) [reply tests/e2e/core-tests/specs/merchant/wp-admin-order-new.test.js]: Done, I already changed the expected currency to USD.

**Comment by rodel-calasagsag** [comment]
Hi @rrennick @tammullen I have updated my code and the "Changes proposed in this Pull Request" section. Please let me know your thoughts and if I should make further changes. Thank you! 😄

**Comment by rrennick** [comment]
> Since the tax classes are unique to this test, you should add them to the products programmatically instead of adding to the default configuration.

@rodel-calasagsag Sorry, I should have explained this in the original comment. You can add the tax class without creating an entirely new product properties object:

```
				const initSimpleProduct = async () => {
					const repo = SimpleProduct.restRepository(httpClient);
					const props = config.get('products.simple');
					return await repo.create( {
						...props,
						taxClass: '<tax class>'
					} );
				};
```

**Comment by rrennick** [comment]
@rodel-calasagsag It was great to see you used the API to create all the base data for this test. It worked great for me locally.
```
Test Suites: 5 passed, 5 total
Tests:       2 skipped, 129 passed, 131 total
Snapshots:   0 total
Time:        569.876s
Ran all test suites.
Jest exit code: 0
```

**Comment by rodel-calasagsag** [comment]
> @rodel-calasagsag Sorry, I should have explained this in the original comment. You can add the tax class without creating an entirely new product properties object:
> 
> ```
> 				const initSimpleProduct = async () => {
> 					const repo = SimpleProduct.restRepository(httpClient);
> 					const props = config.get('products.simple');
> 					return await repo.create( {
> 						...props,
> 						taxClass: '<tax class>'
> 					} );
> 				};
> ```
Hi @rrennick thank you for the review.

I had thought about doing the approach you mentioned above. However, I needed the name and price to be unique to this test. If this test would use the default product name for a simple product, which is "Simple Product", then the test would select the wrong "Simple Product" when creating the order, since there are a number of products with that same name created by other tests.

I also needed the price to be unique and simple so that the tax calculation is easier to verify when debugging. So if I would change the product name and the price, then I would  have changed majority of the properties in the default simple product 😅 , so I constructed an object instead.

Maybe I can do that on the other product types though. Let me double check.

**Comment by rodel-calasagsag** [comment]
Hi @rrennick I got a question by the way. I realized that I used the property `tax_class` instead of `taxClass`. Interestingly, the test passes regardless of which of them I use. So my question is... why? 😄 

I suppose there's some kind of automatic parsing of camelcased and underscored properties that happens somewhere?

For the meantime, I'll change it to `taxClass` to follow the convention.

**Comment by rrennick** [comment]
> I also needed the price to be unique and simple so that the tax calculation is easier to verify when debugging. So if I would change the product name and the price, then I would have changed majority of the properties in the default simple product

That's a good call.

> I suppose there's some kind of automatic parsing of camelcased and underscored properties that happens somewhere?

Yes, the api package has transformers that convert snake case `variable_name` properties to camel case `variableName`. The REST API field is `tax_class` and the transformed field is `taxClass`. When you are submitting data to the api either form works. Responses have camel case properties.

**Comment by rrennick** [comment]
@tammullen This looks great for me.

**Comment by rodel-calasagsag** [comment]
> Yes, the api package has transformers that convert snake case variable_name properties to camel case variableName. The REST API field is tax_class and the transformed field is taxClass. When you are submitting data to the api either form works. Responses have camel case properties.

@rrennick thank you for the clarification!

File: tests/e2e/core-tests/package.json [hunk 2]
```diff
@@ -11,7 +11,8 @@
   "main": "index.js",
   "dependencies": {
     "@jest/globals": "^26.4.2",
-    "config": "3.3.3"
+    "config": "3.3.3",
+    "faker": "^5.5.2"
```
Comment (tammullen) [inline tests/e2e/core-tests/package.json:15]: Is faker used in this test? If not can we remove this addition

**Review by tammullen — CHANGES_REQUESTED** [review]
Looks Great! Thanks for this.

```
 PASS  ../specs/wp-admin/test-create-order.js (33.265s)
  WooCommerce Orders > Add new order
    ✓ can create new order (9795ms)
    ✓ can create new complex order with multiple product types & tax classes (8642ms)

Test Suites: 1 passed, 1 total
Tests:       2 passed, 2 total
Snapshots:   0 total
Time:        33.364s
```
Just have one question about use of faker but after that it looks ready to merge.

File: tests/e2e/core-tests/package.json (hunk 2 above)
Comment (rodel-calasagsag) [reply tests/e2e/core-tests/package.json]: Oh right! Yeah, I just realized this test doesn't require faker anymore. Thanks for spotting this @tammullen 

I removed it from the `/core-tests/package.json` file now

Comment (tammullen) [reply tests/e2e/core-tests/package.json]: Thanks!

**Review by tammullen — APPROVED** [review]

**Pull request merged**